lark-stubs/lexer.pyi

```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+
+from typing import (
+    TypeVar, Type, Tuple, List, Dict, Iterator, Collection, Callable, Optional,
+    Pattern as REPattern,
+)
+from abc import abstractmethod, ABC
+
+_T = TypeVar('_T')
+_MRes = List[Tuple[REPattern, Dict[int, str]]]
+
+
+class Pattern(ABC):
+    value: str
+    flags: Collection[str]
+
+    def __init__(self, value: str, flags: Collection[str] = ...):
+        ...
+
+    @property
+    @abstractmethod
+    def type(self) -> str:
+        ...
+
+    @abstractmethod
+    def to_regexp(self) -> str:
+        ...
+
+    @property
+    @abstractmethod
+    def min_width(self) -> int:
+        ...
+
+    @property
+    @abstractmethod
+    def max_width(self) -> int:
+        ...
+
+
+class PatternStr(Pattern):
+    type: str = ...
+
+    def to_regexp(self) -> str:
+        ...
+
+    @property
+    def min_width(self) -> int:
+        ...
+
+    @property
+    def max_width(self) -> int:
+        ...
+
+
+class PatternRE(Pattern):
+    type: str = ...
+
+    def to_regexp(self) -> str:
+        ...
+
+    @property
+    def min_width(self) -> int:
+        ...
+
+    @property
+    def max_width(self) -> int:
+        ...
+
+
+class TerminalDef:
+    name: str
+    pattern: Pattern
+    priority: int
+
+    def __init__(self, name: str, pattern: Pattern, priority: int = ...):
+        ...
+
+
+class Token(str):
+    type: str
+    pos_in_stream: int
+    value: str
+    line: int
+    column: int
+    end_line: int
+    end_column: int
+    end_pos: int
+
+    def update(self, type_: Optional[str] = None, value: Optional[str] = None) -> Token:
+        ...
+
+    @classmethod
+    def new_borrow_pos(cls: Type[_T], type_: str, value: str, borrow_t: Token) -> _T:
+        ...
+
+
+_Callback = Callable[[Token], Token]
+
+
+def build_mres(terminals: Collection[TerminalDef], match_whole: bool = False) -> _MRes:
+    ...
+
+
+class UnlessCallback:
+    mres: _MRes
+
+    def __init__(self, mres: _MRes):
+        ...
+
+    def __call__(self, t: Token) -> Token:
+        ...
+
+
+class CallChain:
```

Comment:
Why did you add all these internal classes? I don't think 99% of the users will ever even see them..

Reply:
Oh, That's the note when I'm tried to deriving member types in some methods, same as `_Lex`.
I can remove them.

Reply:
The thing is that some of these classes might change in the future, because they aren't part of the API.

How does it help to have these? (For example with _Lex?)

Reply:
I left it behind when assuming the types of some members, because they are generated from some non-public methods.

Reply:
Like `CallChain` and `UnlessCallback` can be simplified as `Callable` but I don't known at first time.